### PR TITLE
[FLINK-20016][python] Support TimestampAssigner and WatermarkGenerator for Python DataStream API.

### PR DESCRIPTION
**Pull request by shuiqiangchen**
## What is the purpose of the change

Support TimestampAssigner and WatermarkGenerator for Python DataStream API.

## Brief change log

- added new interfaces WatermarkStrategy and TimestampAssigner.
- added new interface assign_timestamps_and_watermarks() for DataStream.

## Verifying this change

This change has test case covered by test_timestamp_assigner_and_watermark_strategy() in test_data_stream.py.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? ( not documented)

### CODE REVIEW

File: flink-python/pyflink/datastream/tests/test_data_stream.py [hunk 1]
```diff
@@ -601,6 +602,50 @@ def test_basic_array_type_info(self):
...
+                pass
+
+        watermark_strategy = WatermarkStrategy.for_monotonous_timestamps()\
+            .with_timestamp_assigner(MyTimestampAssigner())
+        data_stream.assign_timestamps_and_watermarks(watermark_strategy)\
+            .key_by(lambda x: x[0], key_type_info=Types.INT()) \
+            .process(MyProcessFunction(), output_type=Types.STRING()).add_sink(self.test_sink)
+        self.env.execute('test time stamp assigner')
+        result = self.test_sink.get_results()
+        expeected_result = ["current timestamp: None, current watermark: 9223372036854775807, "
```
Comment (dianfu) [inline flink-python/pyflink/datastream/tests/test_data_stream.py:637]:
```suggestion
        expected_result = ["current timestamp: None, current watermark: 9223372036854775807, "
```

File: flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+ * events and generate watermarks by user specify TimestampAssigner and WatermarkStrategy.
+ *
+ * <p>These two responsibilities run in the same operator rather than in two different ones,
+ * because the implementation of the timestamp assigner and the watermark generator is
+ * frequently in the same class (and should be run in the same instance), even though the
+ * separate interfaces support the use of different classes.
+ *
+ * @param <IN> The type of the input elements
+ */
+public class TimestampsAndWatermarksOperator<IN> extends StatelessOneInputPythonFunctionOperator<IN, IN>
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java:56]: What about rename to PythonTimestampsAndWatermarksOperator

File: flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+	implements ProcessingTimeCallback {
+
+	private static final long serialVersionUID = 1L;
+
+	private final WatermarkStrategy<IN> watermarkStrategy;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java:64]: Add documentation

File: flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	private transient StreamRecord<IN> reusableStreamRecord;
+
+	public TimestampsAndWatermarksOperator(
+		Configuration config,
+		TypeInformation<IN> inputTypeInfo,
+		DataStreamPythonFunctionInfo pythonFunctionInfo,
+		WatermarkStrategy<IN> watermarkStrategy) {
+		super(config, inputTypeInfo, inputTypeInfo, pythonFunctionInfo);
+		this.watermarkStrategy = watermarkStrategy;
+		this.chainingStrategy = ChainingStrategy.ALWAYS;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java:112]: could be removed as it's already set in the super class

File: flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+		runnerInputSerializer = PythonTypeUtils.TypeInfoToSerializerConverter
+			.typeInfoSerializerConverter(runnerInputTypeInfo);
+		runnerOutputSerializer = PythonTypeUtils.TypeInfoToSerializerConverter
+			.typeInfoSerializerConverter(runnerOutputTypeInfo);
+		resuableInput = new Row(2);
+		this.reusableStreamRecord = new StreamRecord<>(null);
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java:125]:
```suggestion
		reusableStreamRecord = new StreamRecord<>(null);
```
Make the code style consistent.

File: flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+	private transient TypeSerializer runnerOutputSerializer;
+
+	private transient WatermarkGenerator<IN> watermarkGenerator;
+
+	private transient WatermarkOutput watermarkOutput;
+
+	private transient long watermarkInterval;
+
+	private transient Row resuableInput;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java:98]: Typo
```suggestion
	private transient Row reusableInput;
```

File: flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+	public TimestampsAndWatermarksOperator(
+		Configuration config,
+		TypeInformation<IN> inputTypeInfo,
+		DataStreamPythonFunctionInfo pythonFunctionInfo,
+		WatermarkStrategy<IN> watermarkStrategy) {
+		super(config, inputTypeInfo, inputTypeInfo, pythonFunctionInfo);
+		this.watermarkStrategy = watermarkStrategy;
+		this.chainingStrategy = ChainingStrategy.ALWAYS;
+		this.runnerInputTypeInfo = Types.ROW(Types.LONG, inputTypeInfo);
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java:113]: Could move the initialization of runnerInputTypeInfo and runnerOutputTypeInfo to the open method. Then we could make them transient.

File: flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+		byte[] rawResult = resultTuple.f0;
+		int length = resultTuple.f1;
+		bais.setBuffer(rawResult, 0, length);
+		Row runnerOutput = (Row) runnerOutputSerializer.deserialize(baisWrapper);
+		long newTimestamp = (Long) runnerOutput.getField(0);
+		IN originalData = (IN) runnerOutput.getField(1);
+		reusableStreamRecord.replace(originalData, newTimestamp);
+		output.collect(reusableStreamRecord);
+		watermarkGenerator.onEvent(originalData, newTimestamp, watermarkOutput);
+
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java:166]: unnecessary empty line

File: flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+		}
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		watermarkGenerator.onPeriodicEmit(watermarkOutput);
+	}
+
+	private static final class WatermarkEmitter implements WatermarkOutput {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java:211]: could reuse the WatermarkEmitter defined in org.apache.flink.streaming.runtime.operators.WatermarkEmitter.

File: flink-python/pyflink/common/time.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+from pyflink.java_gateway import get_gateway
+
+
+class Duration(object):
```
Comment (dianfu) [inline flink-python/pyflink/common/time.py:21]: Export it

File: flink-python/pyflink/common/watermark_strategy.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+import abc
+
+from pyflink.common.time import Duration
+from pyflink.java_gateway import get_gateway
+
+
+class WatermarkStrategy(object):
```
Comment (dianfu) [inline flink-python/pyflink/common/watermark_strategy.py:24]: export it

File: flink-python/pyflink/common/watermark_strategy.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    """
+    A TimestampAssigner assigns event time timestamps to elements. These timestamps are used by all
+    functions that operate on event time, for example event time windows.
+
+    Timestamps can be an arbitrary int value, but all built-in implementations represent it as the
+    milliseconds since the Epoch (midnight, January 1, 1970 UTC), the same way as time.time() does
+    it.
+    """
+    @abc.abstractmethod
+    def extract_timestamp(self, value, previous) -> int:
```
Comment (dianfu) [inline flink-python/pyflink/common/watermark_strategy.py:98]: add type hint

File: flink-python/pyflink/common/watermark_strategy.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    and the TimestampAssigner which assigns the internal timestamp of a record.
+
+    The convenience methods, for example forBoundedOutOfOrderness(Duration), create a
+    WatermarkStrategy for common built in strategies.
+    """
+    def __init__(self, j_watermark_strategy):
+        self._j_watermark_strategy = j_watermark_strategy
+        self._timestamp_assigner = None
+
+    def with_timestamp_assigner(self, timestamp_assigner: 'TimestampAssigner'):
```
Comment (dianfu) [inline flink-python/pyflink/common/watermark_strategy.py:37]: type hint

File: flink-python/pyflink/common/watermark_strategy.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+    def with_idleness(self, idle_timeout: Duration) -> 'WatermarkStrategy':
+        """
+        Creates a new enriched WatermarkStrategy that also does idleness detection in the created
+        WatermarkGenerator.
+
+        :param idle_timeout: The idle timeout.
+        :return: A new WatermarkStrategy with idle detection configured.
+        """
+        return WatermarkStrategy(self._j_watermark_strategy.withIdleness(idle_timeout._j_duration))
```
Comment (dianfu) [inline flink-python/pyflink/common/watermark_strategy.py:59]: self._timestamp_assigner is missing

File: flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+	/**
+	 * Reusable row for normal data runner inputs.
+	 */
+	private transient Row resuableInput;
+
+	/**
+	 * Reusable StreamRecord for data with new timestamp calculated in TimestampAssigner.
+	 */
+	private transient StreamRecord<IN> reusableStreamRecord;
+
```
Comment (SteNicholas) [inline flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java:104]: This misses the `emitProgressiveWatermarks` flag, which is whether to emit intermediate watermarks or only one final watermark at the end of input.

**Pull request closed without merging**